design/open/reuse-contented-merge-metaranges/reuse-contented-merge-metaranges.md

```diff
@@ -0,0 +1,128 @@
+the destination branch, and _all other merges lose and have to restart_.  So
+on each concurrent merge, every successive merge attempt takes even longer.
+
+This proposal cannot prevent concurrent merges from racing.  Instead it
+makes retries considerably more efficient.  Any range that is affected by
+only one concurrent merge will be written just _once_, no matter how many
+times it is retried.  So additionally retries of the merge that is writing
+that range will be able to skip _reading_ its two sources on every retry.
+This should speed everything up!
+
+### KV or not KV?
+
+Importantly, this proposal is independent of whether or not we use KV.
+_Before_ KV you can use it by removing the branch locker and making the
+branch update operation conditional.  And then it **improves** concurrent
+merge performance by allowing parallel execution of almost all of the
+time-consuming tasks.  So it increases the amount of work performed but it
+performs that work in parallel and reduces the latency of the slow
+concurrent merges.
+
+_After_ KV, not only does it restore performance to this improved case -- it
+also removes much of the duplicate work that concurrent merges perform.
+
+## How
+
+The basic idea is to allow retries of a concurrent merge to re-use previous
+work.  This is similar to the "overlapping `sealed_tokens`" full-commit
+method of the [commit flow][commit_flow].  However here there are no staging
+tokens; the previous work of a merge attempt is the metarange that it
+generated.
+
+![Branches A and B diverge from branch main.  Merge A and B into main concurrently.  A succeeds in merging and bumps HEAD of main.  B succeeds in merging and generates a metarange, but now it cannot bump HEAD of main.  Now try again to merge, but _use the newly-generated merged metarange instead of the HEAD of B.](./diagram.jpg)
+
+There are two ways for a merge to fail:
+* **Conflict:** The objects at some path on the merge base, source and
+  destination are all different and no strategy was specified to resolve it.
+
+  This failure is inherent and occurs regardless of additional concurrent
+  merges.  It occurs while generating the merged metarange.  It cannot be
+  fixed; correct behaviour is simply to report it to the user.
+* **Losing a race:** No merge conflict occurred, but when trying to add the
+  resulting merge commit to the destination we discover that the HEAD of the
+  destination branch has changed.
+
+  This failure is due to losing a race against a concurrent merge or commit.
+  It must either be retried or abandoned (and, presumably, retried later).
+
+Naturally only race failures are of interest.  When a merge fails because of
+a merge, it must be retried.  However we can restart it using _the generated
+merged metarange_ as source and the original destination as the base!  (See
+"Correctness" below for why this yields the correct result.)
+
+The performance gain occurs during the retry: We expect the new source to
```

Comment:
To gain performance improvement the assumption here is that the two merges don't change the same ranges, need to make sure that this is the case

Reply:
You're right: this will be true for some but not all merges.  With our emerging focus on merging as the basic tool for handling multi-object file formats (Spark outputs, probably Iceberg, maybe Delta too), I would expect ranges to separate if there are repeated racing merges -- simply because large multi-objects will split off into their range soon enough :-)

Adding words to that effect.